### ISSUE TITLE
usage about Pipfile.lock on multiple paltforms(window, linux, macos)

### ISSUE THREAD

**Issue by leileigong** [pypa/pipenv#3702]
I have a ython application runs on multiple paltforms(window, linux, macos).
I only wanna to manage one  `Pipfile `and add it to git control. 

but the Pipfile.lock generated by `pipenv install ` may be different on different platforms.
So it seems that `Pipfile.lock
` should not be controlled by git!

for example:
Pipfile content:
```
[[source]]
url = "https://pypi.org/simple"
verify_ssl = true
name = "pypi"

[packages]
pyinstaller = {version = "*",markers = """python_version >= '2.7' and python_version != '3.0'and python_version != '3.1' and python_version != '3.2' and python_version != '3.3'"""}

[dev-packages]

[requires]
python_version = "3.6"
```

use `pipenv graph
` insight into my dependency.
> pipenv graph

dependency on Windows:
```
PyInstaller==3.4
  - altgraph [required: Any, installed: 0.16.1]
  - macholib [required: >=1.8, installed: 1.11]
    - altgraph [required: >=0.15, installed: 0.16.1]
  - pefile [required: >=2017.8.1, installed: 2019.4.14]
    - future [required: Any, installed: 0.17.1]
  **- pywin32-ctypes [required: Any, installed: 0.2.0]**
  **- setuptools [required: Any, installed: 40.8.0]**
```
dependency on linux:
```
PyInstaller==3.4
  - altgraph [required: Any, installed: 0.16.1]
  - macholib [required: >=1.8, installed: 1.11]
    - altgraph [required: >=0.15, installed: 0.16.1]
  - pefile [required: >=2017.8.1, installed: 2019.4.14]
    - future [required: Any, installed: 0.17.1]
  - setuptools [required: Any, installed: 41.0.0]
```

So the Pipfile.locks  on windows and linux are different.
How could i solve it?

**Comment by frostming** [comment]
True, I am also facing this problem which is annoying. The only way to work is to write the missing packages to Pipfile with system markers.

The problem is the resolver can't know all dependencies when the resolution is performed on a particular platform, which makes the whole environment only reproducible on the same platform. I will mark it as an enhancement for future.

**Comment by leileigong** [comment]
can pipenv generate different Pipfile.lock with specified file name? such as Pipfile_win.lock for windows paltform, Pipfile_mac.lock for macos paltform.

**Comment by lig** [comment]
There is also a similar use case for gpu/non-gpu systems even within the same platform and I'm not sure if it fits here. At the moment, we install gpu targeted packages manually over the installation from Pipefile.lock. There are a lot of scientific packages which have different builds depending on whether you want it to use gpu or not. So, we end up with cpu-targeted container image and a hacky one gpu-targeted.

**Comment by thomasf** [comment]
> can pipenv generate different Pipfile.lock with specified file name? such as Pipfile_win.lock for windows paltform, Pipfile_mac.lock for macos paltform.

The way pipenv insists on selecting the latest versions all the time kind of makes that solution bad. If the lock file isn't updated for all platforms at the same time it's too likely that developers for different platforms will be sitting with different versions rendering the whole idea of using Pipenv in the first place questionable. 

I have btw. suggested support for minimal version selection in pipenv which might help increase predictability of installs https://github.com/pypa/pipenv/issues/3701 even though there are still no guarantees when dependencies are selected dynamically during setup.

**Comment by thomasf** [comment]
I have only written python packages which are simple dependencies which can be declared as a list in `setup.py` so I'm not really aware of what the detailed problems are here..  The only reliable way I could see this working is if Pipenv can know that all dependencies are fully declared in `setup.cfg` . Something along the lines of a named "gpu/nogpu" section like the `extras_require` section works ( https://setuptools.readthedocs.io/en/latest/setuptools.html#declaring-extras ) which could then be selected during install.  This would allow pipenv to also record the gpu/nogpu sections into it's lock file and understand how to merge/update/read parts of the lock file..
